sys/kern/dev_procstat.c

```diff
@@ -0,0 +1,130 @@
+#include <sys/devfs.h>
+#include <sys/vnode.h>
+#include <sys/linker_set.h>
+#include <sys/queue.h>
+#include <sys/proc.h>
+#include <sys/thread.h>
+#include <sys/klog.h>
+#include <sys/malloc.h>
+#include <stdio.h>
+
+/* Implementation of /dev/procstat
+ *
+ * When /dev/procstat is opened it takes a snapshot of info of all existing
+ * processes.
+ *
+ * During read calls to /dev/procstat this info can be read.
+ *
+ * Example:
+ * euid   pid    ppid    pgrp   session  state    td_name  elfpath
+ * 0       1       0       1       1       R       init    /bin/ksh
+ *
+ */
+
+#define PSTAT_SIZE 1024
+#define BUF_SIZE (PSTAT_SIZE - sizeof(size_t))
+
+static char state[4] = {
+  [PS_NORMAL] = 'R',
+  [PS_STOPPED] = 'S',
+  [PS_DYING] = 'D',
+  [PS_ZOMBIE] = 'Z',
+};
+
+typedef struct pstat_data {
+  size_t len;
+  char buf[];
+} pstat_t;
+
+static int dev_procstat_read(file_t *f, uio_t *uio) {
+  pstat_t *pst = f->f_data;
+
+  if ((size_t)f->f_offset >= pst->len)
+    return 0;
+
+  uio->uio_offset = f->f_offset;
+  size_t len = MIN(uio->uio_resid, pst->len);
```

Comment:
Shouldn't it be more like `MIN(uio->uio_resid, pst->len - uio->uio_offset);` ?

Reply:
Yes. Thanks!